capsules/src/lora/radio.rs

```diff
@@ -0,0 +1,761 @@
+            //self.wake_pending.set(true);
+        }
+
+        ReturnCode::SUCCESS
+    }
+
+    fn stop(&self) -> ReturnCode {
+        if self.state.get() == InternalState::SLEEP || self.state.get() == InternalState::TX_OFF
+        {
+            return ReturnCode::EALREADY;
+        }
+
+        match self.state.get() {
+            InternalState::READY => {
+                self.state.set(InternalState::START);
+            }
+            _ => {
+                self.sleep_pending.set(true);
+            }
+        }
+
+        ReturnCode::SUCCESS
+    }
+}
+
+// The modem
+pub struct Radio<'a, S: SpiMasterDevice> {
+  spi: &'a S,
+  spi_buf: TakeCell<'static, [u8]>,
+  spi_rx: TakeCell<'static, [u8]>,
+  spi_tx: TakeCell<'static, [u8]>,
+  spi_busy: Cell<bool>,
+  //Pins
+  //cs_pin: &'a dyn gpio::Pin,
+  reset_pin: &'a dyn gpio::Pin,
+  irq_pin: &'a dyn gpio::InterruptPin,
+  //State params
+  sleep_pending: Cell<bool>,
+  wake_pending: Cell<bool>,
+  interrupt_handling: Cell<bool>,
+  interrupt_pending: Cell<bool>,
+  state: Cell<InternalState>,
+  //LoRa params
+  frequency: Cell<u64>,
+  packet_index: Cell<u8>,
+  implicit_header: Cell<bool>,
+  tx_done: bool,
+  rx_done: bool,
+}
+
+impl<S: SpiMasterDevice> Radio<'a, S> {
+  pub fn new(
+      spi: &'a S,
+      //cs: &'a dyn gpio::Pin,
+      reset: &'a dyn gpio::Pin,
+      irq: &'a dyn gpio::InterruptPin,
+  ) -> Radio<'a, S> {
+      Radio {
+          spi: spi,
+          spi_buf: TakeCell::empty(),
+          spi_rx: TakeCell::empty(),
+          spi_tx: TakeCell::empty(),
+          spi_busy: Cell::new(false),
+          //cs_pin: cs,
+          reset_pin: reset,
+          irq_pin: irq,
+          state: Cell::new(InternalState::START),
+          sleep_pending: Cell::new(false),
+          wake_pending: Cell::new(false),
+          interrupt_handling: Cell::new(false),
+          interrupt_pending: Cell::new(false),
+          frequency: Cell::new(0),
+          packet_index: Cell::new(0),
+          implicit_header: Cell::new(false),
+          tx_done: false,
+          rx_done: false,
+      }
+  }
+
+  // SPI operations for handling LoRa IRQ
+  fn handle_interrupt(&self) {
+      if self.spi_busy.get() == false {
+          if self.state.get() == InternalState::RX_ON {
+              // We've received a complete frame; need to disable
+              // reception until we've read it out from RAM,
+              // otherwise subsequent packets may corrupt it.
+             self.state.set(InternalState::RX_OFF);
+          } else {
+              self.interrupt_handling.set(true);
+          }
+      } else {
+          self.interrupt_pending.set(true);
+      }
+  }
+    
+  fn register_write(&self, reg: RegMap, val: u8) -> ReturnCode {
+      //if self.spi_busy.get() || self.spi_tx.is_none() || self.spi_rx.is_none() {
+      //    return ReturnCode::EBUSY;
+      //}
+      let wbuf = self.spi_tx.take().unwrap();
+      let rbuf = self.spi_rx.take().unwrap();
```

Comment:
I think this can be:
```rust
self.spi_tx.take().map(|wbuf| {
    self.spi_rx.take().map(|rbuf| { ... }
)});
```

Reply:
Any advantage of this particular syntax? I find the current one much cleaner and also existing elsewhere in Tock.

Reply:
Using `.map` automatically puts the buffers back into the cell after the closure execution. Using `.take()` will probably make you forget that at some point. `.take()` is used elsewhere when the buffers are passed to components returning them in a callback, as the references in `.map()` will only be valid for the closure and therefore can't be moved out of it.